File: julien-lectures/CODE/Program_7_2.c
/* ****************************************************************

This is the C version of program 7.2 from page 242 of
"Modeling Infectious Disease in humans and animals"
by Keeling & Rohani.
 
It is the simple SIR epidemic in a metapopulation with human movement patterns. 
For simplicity births and deaths have been ignored, and we work with numbers of individuals.
Y[i][j] refers to infected individual who are currently in i but live in j.
For simplicity we set MAX to be the largest number of subpopulations it is 
possible to have -- this can easily be altered.

This code is written to be simple, transparent and readily compiled.
Far more elegant and efficient code can be written.

This code can be compiled using gnu or intel C compilers:
icc -o Program_7_1  Program_7_1.c  -lm
g++ -o Program_7_1  Program_7_1.c  -lm

If you have any difficulties, we suggest you set DEBUG to 1 or -1

******************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

// Maximum number of subpopulations
#define MAX 20

// Set DEBUG to 1 to output some debugging information
// Set DEBUG to -1 to output more debugging information
#define DEBUG 1

// Set up basic parameters
int N_subpops;
double beta[MAX];
double gamm[MAX];
double X0[MAX];
double Y0[MAX];
double N0[MAX];

double l[MAX][MAX];
double r[MAX][MAX];

double MaxTime;

// Set up variables and rates of change
double t,X[MAX][MAX],Y[MAX][MAX],N[MAX][MAX],Pop[3*MAX*MAX];
double dPop[3*MAX*MAX];

// Initialise the equations and Runge-Kutta integration
void Diff(double Pop[3*MAX*MAX])
{
  
  int i,j,k,K,h,H;
  // Set up temporary variables to make the equations look neater
  double tmpX, tmpY, tmpN, sumY[MAX], sumN[MAX];


  //calculate number currently in subpopulation i
  for(i=0;i<N_subpops;i++)
    {
      sumN[i]=0; sumY[i]=0;
      for(j=0;j<N_subpops;j++)
	{
	  k=3*(j+i*N_subpops);
	  sumN[i]+=Pop[2+k];
	  sumY[i]+=Pop[1+k];
	}
    }


  // set all rates of change to zero
  for(i=0;i<N_subpops;i++)
    {
      for(j=0;j<N_subpops;j++)
	{
	  k=3*(j+i*N_subpops);
	  dPop[k]=0; dPop[1+k]=0; dPop[2+k]=0;
	}
    } 

  // now calculate the rates of change
  for(i=0;i<N_subpops;i++)
    {
      for(j=0;j<N_subpops;j++)
	{
	  k=3*(j+i*N_subpops); K=3*(i+j*N_subpops); 
	  h=3*(i+i*N_subpops); H=3*(j+j*N_subpops);
	  
	  dPop[k]-=beta[i]*Pop[k]*sumY[i]/sumN[i];    // Infection
	  dPop[k+1]+=beta[i]*Pop[k]*sumY[i]/sumN[i];  // Infection
	  dPop[k+1]-=gamm[i]*Pop[k+1];          // Recovery

	  /* Movement */
	 
	  dPop[h]+=r[j][i]*Pop[K];
	  dPop[h]-=l[j][i]*Pop[h];
	  
	  dPop[1+h]+=r[j][i]*Pop[1+K];
	  dPop[1+h]-=l[j][i]*Pop[1+h];
	  
	  dPop[2+h]+=r[j][i]*Pop[2+K];
	  dPop[2+h]-=l[j][i]*Pop[2+h];
	  
	  dPop[k]+=l[i][j]*Pop[H];
	  dPop[k]-=r[i][j]*Pop[k];
	  
	  dPop[k+1]+=l[i][j]*Pop[1+H];
	  dPop[k+1]-=r[i][j]*Pop[1+k];
	  
	  dPop[k+2]+=l[i][j]*Pop[2+H];
	  dPop[k+2]-=r[i][j]*Pop[2+k];
	    
	}
    }

  return;
}

void Runge_Kutta(double);
void Read_in_parameters(FILE *);
void Perform_Checks();
void Output_Data(FILE *);


// The main program

int main(int argc, char** argv)
{
  char FileName[1000];
  int i,j;
  double step, Every, tmp;
  double TotalS, TotalI;
  FILE *in, *out;
   
  /* Find the given parameter  file name and open the file 
     otherwise use parameters set at top of program */
  if(argc>1)
    {
      strcpy(FileName, *(argv+1));  
      
      if((in = fopen(FileName,"r"))==NULL)
	{
	  printf("Cannot read file %s\n",FileName);
	  exit(1);
	}

      /* Read in the parameters */
      
      Read_in_parameters(in);

      if(DEBUG) printf("\nReading parameters from file %s\n",FileName);
    }
  else
    {
      if(DEBUG) printf("\nSetting default parameters\n");
	
      N_subpops=5;
      for(i=0;i<N_subpops;i++)
	{
	  beta[i]=1;
	  gamm[i]=0.3;
	  for(j=0;j<N_subpops;j++)
	    {
	      if( abs(i-j)==1) l[i][j]=0.1;
	      else l[i][j]=0;
	      r[i][j]=2.0;
	    }
	  r[i][i]=0;
	  N0[i]=1000;
	  X0[i]=800;
	  Y0[i]=0;
	}
      Y0[0]=1;
      MaxTime=60;
    }

  /* Check all parameters are OK & set up intitial conditions */

  Perform_Checks();
  for(i=0;i<N_subpops;i++)
     {
       for(j=0;j<N_subpops;j++)
	 {
	   X[i][j]=0; Y[i][j]=0; N[i][j]=0;
	 }
       X[i][i]=X0[i]; Y[i][i]=Y0[i]; N[i][i]=N0[i];
     }


   /* Find a suitable time-scale for outputs */

   step=1e100;
   for(i=0;i<N_subpops;i++)
     {
       tmp=0;
       for(j=0;j<N_subpops;j++)
	 {
	   tmp+=r[i][j]+l[i][j];
	 }
       tmp=0.1/((beta[i]+gamm[i]+tmp)*X0[i]);

       if(tmp<step)
	 {
	   step=tmp;
	   Every=pow(10,floor(log10(100.0*tmp)));
	 }
     }
  while(MaxTime/Every>10000)
    {
      Every*=10.0;
    }
  

  if(DEBUG) printf("Using a time step of %g and outputing data every %g\n\n",step,Every);
 


  if((out = fopen("Output","w"))==NULL)
    {
      printf("Cannot open file Output to write data\n");
      exit(1);
    }

  /* The main iteration routine */

  t=0;

  Output_Data(out);

  do
    {
      Runge_Kutta(step);  t+=step;
      
      /* If time has moved on sufficiently, output the current data */
      if( floor(t/Every) > floor((t-step)/Every) )
	{
	  Output_Data(out);
	}
      else
	{
	  if(DEBUG<0) Output_Data(stdout);
	}
      
    }
  while(t<MaxTime);

  Output_Data(out);

  fclose(out);
}



void Read_in_parameters(FILE *in)
{
  int i,j;
  char str[200];

  fscanf(in,"%s",str);
  fscanf(in,"%d",&N_subpops);
  if(N_subpops>MAX) {printf("Number of subpopulations N_subpops (%d) exceeds the predefined maximum\n",N_subpops); exit(1);}
  
  fscanf(in,"%s",str);
  for(i=0;i<N_subpops;i++)
    {
      fscanf(in,"%lf",&beta[i]);
    }
  
  fscanf(in,"%s",str);
  for(i=0;i<N_subpops;i++)
    {
      fscanf(in,"%lf",&gamm[i]);
    }
  
  fscanf(in,"%s",str);
  for(i=0;i<N_subpops;i++)
    {
      for(j=0;j<N_subpops;j++)
	{
	  fscanf(in,"%lf",&l[i][j]);
	}
    }
  
  fscanf(in,"%s",str);
  for(i=0;i<N_subpops;i++)
    {
      for(j=0;j<N_subpops;j++)
	{
	  fscanf(in,"%lf",&r[i][j]);
	}
    }
  
  fscanf(in,"%s",str);
  for(i=0;i<N_subpops;i++)
    {
      fscanf(in,"%lf",&N0[i]);
    }
  
  fscanf(in,"%s",str);
  for(i=0;i<N_subpops;i++)
    {
      fscanf(in,"%lf",&X0[i]);
    }
  
  fscanf(in,"%s",str);
  for(i=0;i<N_subpops;i++)
    {
      fscanf(in,"%lf",&Y0[i]);
    }
  
  fscanf(in,"%s",str);
  fscanf(in,"%lf",&MaxTime);
  
  fclose(in);
  
}


void Perform_Checks()
{
  int i,j;

  for(i=0;i<N_subpops;i++)
    {
      if(X0[i]<=0) {printf("ERROR: Initial level of susceptibles_%d (%g) is less than or equal to zero\n",i,X0[i]); exit(1);}
      
      if(Y0[i]<0) {printf("ERROR: Initial level of infecteds_%d (%g) is less than zero\n",i,Y0[i]); exit(1);}
      
      if(beta[i]<=0) {printf("ERROR: Transmission rate beta_%d (%g) is less than or equal to zero\n",i,beta[i]); exit(1);}
      
      if(gamm[i]<=0) {printf("ERROR: Recovery rate gamma_%d (%g) is less than or equal to zero\n",i,gamm[i]); exit(1);}
      
      if(X0[i]+Y0[i]>N0[i]) {printf("WARNING: Initial level of susceptibles+infecteds (%g+%g=%g) is greater than population size (%g)\n",X0[i],Y0[i],X0[i]+Y0[i],N0[i]);}
      
      if(beta[i]<gamm[i]) {printf("WARNING: Basic reproductive ratio (R_0=%g) is less than one\n",beta[i]/(gamm[i]));}
      
      for(j=0;j<N_subpops;j++)
	{
	  if(l[i][j]<0) {printf("ERROR: Movement rate l_%d,%d (%g) is less than zero\n",i,j,l[i][j]); exit(1);}
	  if(l[i][j]<0) {printf("ERROR: Return rate r_%d,%d (%g) is less than zero\n",i,j,r[i][j]); exit(1);}
	}
      if(l[i][i]!=0) {printf("WARNING: Movement rate l_%d,%d (%g) is not equal to zero\n",i,i,l[i][i]);}
      if(r[i][i]!=0) {printf("WARNING: Movement rate r_%d,%d (%g) is not equal to zero\n",i,i,r[i][i]);}
    }
  if(MaxTime<=0) {printf("ERROR: Maximum run time (%g) is less than or equal to zero\n",MaxTime); exit(1);}
}


void Output_Data(FILE *out)
{
  int i,j;
  double TotalS, TotalI;

  if(out!=stdout)
    {
      fprintf(out,"%g \t",t);
      for(i=0;i<N_subpops;i++)
	{
	  TotalS=TotalI=0;
	  for(j=0;j<N_subpops;j++)
	    {
	      TotalS+=X[i][j]; TotalI+=Y[i][j];
	    }
	  fprintf(out," %g %g",TotalS,TotalI);
	}
      fprintf(out,"\n");
    }

  if(DEBUG)
    {
      TotalS=TotalI=0;
      for(i=0;i<N_subpops;i++)
	{
	  for(j=0;j<N_subpops;j++)
	    {
	      TotalS+=X[i][j]; TotalI+=Y[i][j];
	    }
	}
      printf("%g \t %g %g\n",t,TotalS,TotalI);
    }


  if(DEBUG<0)
    {
      printf("%g \t",t);
      for(i=0;i<N_subpops;i++)
	{
	  TotalS=TotalI=0;
	  for(j=0;j<N_subpops;j++)
	    {
	      TotalS+=X[i][j]; TotalI+=Y[i][j];
	    }
	  printf(" %g %g",TotalS,TotalI);
	}
      printf("\n");
    }

}





void Runge_Kutta(double step)
{
  int i,j,k;
  double dPop1[3*MAX*MAX], dPop2[3*MAX*MAX], dPop3[3*MAX*MAX], dPop4[3*MAX*MAX];
  double tmpPop[3*MAX*MAX], initialPop[3*MAX*MAX];

  /* Integrates the equations one step, using Runge-Kutta 4
     Note: we work with arrays rather than variables to make the
     coding easier */

  for(i=0;i<N_subpops;i++)
    {
      for(j=0;j<N_subpops;j++)
	{
	  k=3*(j+i*N_subpops);
	  initialPop[k]=X[i][j]; initialPop[1+k]=Y[i][j]; initialPop[2+k]=N[i][j];
	}
    }
  
  Diff(initialPop);
  for(i=0;i<3*N_subpops*N_subpops;i++)
    {
      dPop1[i]=dPop[i];
      tmpPop[i]=initialPop[i]+step*dPop1[i]/2;
    }

  Diff(tmpPop);
  for(i=0;i<3*N_subpops*N_subpops;i++)
    {
      dPop2[i]=dPop[i];
      tmpPop[i]=initialPop[i]+step*dPop2[i]/2;  
    }

  Diff(tmpPop);
  for(i=0;i<3*N_subpops*N_subpops;i++)
    {
      dPop3[i]=dPop[i];
      tmpPop[i]=initialPop[i]+step*dPop3[i]; 
    }

  Diff(tmpPop);

  for(i=0;i<3*N_subpops*N_subpops;i++)
    {
      dPop4[i]=dPop[i];

      tmpPop[i]=initialPop[i]+(dPop1[i]/6 + dPop2[i]/3 + dPop3[i]/3 + dPop4[i]/6)*step;
    }

  for(i=0;i<N_subpops;i++)
    {
      for(j=0;j<N_subpops;j++)
	{
	  k=3*(j+i*N_subpops);
	  X[i][j]=tmpPop[k]; Y[i][j]=tmpPop[1+k]; N[i][j]=tmpPop[2+k];
	}
    }
  
  return;
}

File: julien-lectures/CODE/Program_7_2.m
function [t,X,Y] = Program_7_2(n,beta,gamma,l,r,N0,X0,Y0,MaxTime)
%
% 
%
% Program_7_2( n,beta, gamma, l, r, N0, X0, Y0, MaxTime)
%      This is the MATLAB version of program 7.2 from page 242 of 
% "Modeling Infectious Disease in humans and animals" 
% by Keeling & Rohani.
% 
% It is the SIR epidemic in a metapopulationFor simplicity births and deaths have been 
% ignored, and we work with numbers of individuals.
% Y[i][j] refers to infected individual who are currently in i but live in j..
%

% Sets up default parameters if necessary.
if nargin == 0
   n=5;
   beta=1.0*ones(n,1);
   gamma=0.3*ones(n,1);
   N0=1000*ones(n,1);
   X0=800*ones(n,1);
   Y0=0.0*ones(n,1); Y0(1)=1;
   MaxTime=60;
   l=0.1*(diag(ones(1,n-1),1)+diag(ones(1,n-1),-1));
   r=2*ones(n,n); r=r-diag(diag(r));
end

% Check all parameters are the right size
beta=CheckSize(beta,n,1,'beta');
gamma=CheckSize(gamma,n,1,'gamma');
l=CheckSize(l,n,n,'m');
r=CheckSize(r,n,n,'m');
N0=CheckSize(N0,n,1,'S0');
X0=CheckSize(X0,n,1,'S0');
Y0=CheckSize(Y0,n,1,'I0');

% Checks all the parameters are valid
CheckGreaterOrEqual(beta,0,'beta');
CheckGreaterOrEqual(gamma,0,'gamma');
CheckGreaterOrEqual(l,0,'m');
CheckGreaterOrEqual(r,0,'m');
CheckGreaterOrEqual(N0,0,'S0');
CheckGreaterOrEqual(X0,0,'S0');
CheckGreaterOrEqual(Y0,0,'I0');

if sum(abs(diag(l)))>0
    warning('Diagonal terms in movement matrix l are non-zero');
end
if sum(abs(diag(r)))>0
    warning('Diagonal terms in return rate matrix r are non-zero');
end
m=find(X0+Y0>N0);
if length(m)>0
    error('Population %g, susceptibles + infecteds are greater than population size',m(1));
end

X=zeros(n,n); Y=zeros(n,n); NN=zeros(n,n);
X=diag(X0); Y=diag(Y0); NN=diag(N0); All=reshape([X Y NN],3*n*n,1);

% The main iteration 
options = odeset('RelTol', 1e-4);
[t, pop]=ode45(@Diff_7_1,[0 MaxTime],All,options,n, beta, gamma, l, r);

X=zeros(length(t),n); Y=X;
for i=1:n
    X(:,i)=sum(pop(:,i:n:(n*n))')';
    Y(:,i)=sum(pop(:,(n*n)+[i:n:(n*n)])')';
end

% plots the graphs with scaled colours
subplot(2,1,1)
h=plot(t,X,'-g');
for i=1:n
    set(h(i),'Color',[0 0.5+0.5*(i-1)/(n-1) 0]);
end
legend(h);
xlabel 'Time';
ylabel 'Susceptibles'

subplot(2,1,2) 
h=plot(t,Y,'-r');
for i=1:n
    set(h(i),'Color',[0.5+0.5*(i-1)/(n-1) 0 0]);
end
legend(h);
xlabel 'Time';
ylabel 'Infectious'



% Calculates the differential rates used in the integration.
function dPop=Diff_7_1(t,pop, n, beta, gamma, l, r)

X=reshape(pop(1:(n*n)),n,n); Y=reshape(pop((n*n)+[1:(n*n)]),n,n); 
NN=reshape(pop(2*(n*n)+[1:(n*n)]),n,n);

dX=zeros(n,n); dY=zeros(n,n); dNN=zeros(n,n);

% Note the different use of .* and *

sumY=sum(Y')'; sumNN=sum(NN')';

% First do the off diagonals.
dX = -X.*((beta.*sumY./sumNN)*ones(1,n)) + l.*(ones(n,1)*diag(X)')   - r.*X;
dY = X.*((beta.*sumY./sumNN)*ones(1,n)) - Y.*(gamma*ones(1,n)) + l.*(ones(n,1)*diag(Y)')   - r.*Y;
dNN =  l.*(ones(n,1)*diag(NN)')   - r.*NN;

% Now do diagonals
DX = -diag(X).*(beta.*sumY./sumNN) - sum(l)'.*diag(X) + sum(r.*X)';
DY = diag(X).*(beta.*sumY./sumNN) - diag(Y).*gamma - sum(l)'.*diag(Y) + sum(r.*Y)';
DNN = - sum(l)'.*diag(NN) + sum(r.*NN)';

dX = dX - diag(diag(dX)) + diag(DX);
dY = dY - diag(diag(dY)) + diag(DY);
dNN = dNN - diag(diag(dNN)) + diag(DNN);

dPop=reshape([dX dY dNN],3*n*n,1);

% Does a simple check on the size and possible transpose
function Parameter=CheckSize(Parameter, L, W, str)

[l w]=size(Parameter);

if(l==W & w==L)
    Parameter=Parameter'; [l w]=size(Parameter);
end

if(l==1 & w==1)
    warning('Parameter %s is a scaler value, expanding to size %d x %d',str,L,W);
    Parameter=Parameter*ones(L,W);
    [l w]=size(Parameter);
end

if(l~=L | w~=W)
    error('Parameter %s is of size %d x %d and not %d x %d',str,l,w,L,W);
end


% Does a simple check on the value
function []=CheckGreaterOrEqual(Parameter, Value, str)

m=find(Parameter<Value);
if length(m)>0
    error('Parameter %s(%g) (=%g) is less than %g',str,m(1),Parameter(m(1)),Value);
end


